File: leetcodes/5. Longest Palindromic Substring/longestPalindromicSubstring.java
class Solution {
  public String longestPalindrome(String s) {
    //input string, output: string
    //time O(n^2), space O(1)
    //use left and right iterate over the string with the same index
      //char c = cur char of left and right
      //while left in boundary and cur char is c, left --
      //while right in boundary and cur char is c, right ++
      //while left, right in boundary
        //if cur char is not the same, break
        //left --, right ++
      //left = left + 1
      //if end - start < right - left => start = left, end = right
    //return s.substring(start, end)
    int start = 0;
    int end = 0;
    for (int i = 0; i < s.length(); i++) {
      char c = s.charAt(i);
      int left = i;
      int right = i;
      while(left >=0 && s.charAt(left) == c){
        left --;
      }
      while(right<s.length() && s.charAt(right) == c) {
        right++;
      }
      while(left >=0 && right < s.length()) {
        if (s.charAt(left) != s.charAt(right)) break;
        left --;
        right ++;
      }
      left = left + 1;
      if (end - start < right - left) {
        end = right;
        start = left;
      }
    }
    return s.substring(start, end);
  }
}

File: leetcodes/5. Longest Palindromic Substring/longestPalindromicSubstring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
       #two pointers checks from (i, i) start from center the sides to check if it's palindrome
       #two pointers check from (i, i+1) -> consider repeating characters in the row
       #time O(n^2) space O(1)
       longest = ''
       def findLongest(s,l,r):
         while l>=0 and r < len(s) and s[l] == s[r]:
           l -= 1
           r += 1
         return s[l+1:r]

       for i in range(len(s)):
         s1 = findLongest(s, i, i)
         if len(s1) > len(longest): longest = s1
         s2 = findLongest(s, i, i + 1)
         if len(s2) > len(longest): longest = s2
       return longest

       #DP -> use a 2D matrix to store  the start and end index n^2 time O(n^2), space O(n^2)
